cpp/include/graph.hpp

```diff
@@ -0,0 +1,218 @@
+    number_of_edges(number_of_edges_)
+  {}
+};
+
+/**
+ * @Synopsis    A graph stored in CSR (Compressed Sparse Row) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename VT = int, typename WT = float>
+class GraphCSR {
+public:
+  VT const *offsets;       ///< CSR offsets
+  VT const *indices;       ///< CSR indices
+  WT const *edge_data;     ///< edge weight
+
+  GraphProperties          prop;
+
+  size_t                   number_of_vertices;
+  size_t                   number_of_edges;
+
+  /**
+   * @Synopsis    Fill the identifiers array with the vertex identifiers.
+   *
+   * @param[out]    identifier      Pointer to device memory to store the vertex identifiers
+   */
+  void get_vertex_identifiers(VT *identifiers) const {
+    CUGRAPH_EXPECTS( offsets != nullptr , "No graph specified");
+    cugraph::detail::sequence<VT>(number_of_vertices, identifiers);
+  }
+  
+  /**
+   * @Synopsis    Fill the identifiers in the array with the source vertex identifiers
+   *
+   * @param[out]    src_indices      Pointer to device memory to store the source vertex identifiers
+   */
+  void get_source_indices(VT *src_indices) const {
+    CUGRAPH_EXPECTS( offsets != nullptr , "No graph specified");
+    cugraph::detail::offsets_to_indices<VT>(offsets, number_of_vertices+1, src_indices);
+  }
+
+  /**
+   * @Synopsis   Wrap existing arrays representing adjacency lists in a Graph.
+   *             GraphCSR does not own the memory used to represent this graph. This
+   *             function does not allocate memory.
+   *
+   * @Param  offsets               This array of size V+1 (V is number of vertices) contains the offset of adjacency lists of every vertex.
+   *                               Offsets must be in the range [0, E] (number of edges).
+   * @Param  indices               This array of size E contains the index of the destination for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @Param  edge_data             This array of size E (number of edges) contains the weight for each edge.  This
+   *                               array can be null in which case the graph is considered unweighted.
+   * @Param  number_of_vertices    The number of vertices in the graph
+   * @Param  number_of_edges       The number of edges in the graph
+   */
+  GraphCSR(VT const *offsets_, VT const *indices_, WT const *edge_data_,
+           size_t number_of_vertices_, size_t number_of_edges_):
+    offsets(offsets_),
+    indices(indices_),
+    prop(),
+    edge_data(edge_data_),
+    number_of_vertices(number_of_vertices_),
+    number_of_edges(number_of_edges_)
+  {}
+};
+
+/**
+ * @Synopsis    A graph stored in CSC (Compressed Sparse Column) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename VT = int, typename WT = float>
```

Comment:
I like the idea of having a GraphCSC primarily for doing compile time type checking.  Let me see if I can do something to reduce the code duplication here.

Reply:
Just pushed an update with a new pair of base classes that clean up the code replication issue, I think.

Reply:
We could have a `transposed` flag in the CSR structure to indicate that it contains the transposed view. The typedef trick could redirect to the CSR constructor with the transposed flag on.

Reply:
https://github.com/rapidsai/cugraph/pull/722/commits/09adaa80b619c811ced0ab08f977237aebca6c5d is a very good tradeoff

Reply:
Having a transposed flag sounds like a good idea, or we can add a template parameter
```
template <bool transposed = false>
```
If this parameter will be accessed in performance critical parts of the code (I don't think so, but just in case).